content/blog/2020-12-22-pipeline-II-analysis/index.md

```diff
@@ -0,0 +1,150 @@
+given by the formula (1):
+<img src="https://github.com/Sibylau/cs6120/blob/project-jl3952/content/blog/2020-12-22-pipeline-II-analysis/mii.PNG" width="300" />
+
+Where R is the set of available resources. $O_r$ is the number of operations in the loop 
+body which occupy the resource $r$, and $N_r$ is the number of allocated resources, for 
+example, the number of memory ports, or the number of DSPs. 
+
+RecMII stands for Recurrence minimum II. It denotes the lower-bound of II due to loop-carried 
+dependence. A loop-carried dependence indicates that operations in the subsequent loop 
+iteration cannot start until the operations in the current iteration have finished. Array 
+accesses are a common source of loop-carried dependence. For example, the following code 
+snippet shows that the next iteration of the loop will read the array element updated by the 
+current loop iteration. The minimum initiation interval, in this case, is the total number of 
+clock cycles required for memory read, the add operation, and the memory write.
+
+```c
+for (i = 1; i < N; i++)
+    mem[i] = mem[i-1] + i;
+```
+
+The overall minimum initiation interval, considering both resource constraints and data 
+dependences, is $MII = Max(ResMII, RecMII)$, which takes the maximum achievable II under both 
+circumstances. We can take a look at a warm-up example:
+
+```c
+int test()
+{
+  int A[20];
+  for (unsigned t = 2; t <= 20 - 1; t++)
+        A[t] = A[t - 2] + A[t - 1];
+  return 0;
+}
+```
+
+There are two load operations and one write operation in the loop body. If we assume the array 
+`A` is mapped to a RAM with one read port and one write port, then we can calculate ResMII to 
+be 2, as the number of memory read operations divided by the number of read ports. As for 
+loop-carried dependency, we notice that the current loop iteration needs to read array elements
+computed by the previous loop (`A[t - 1]`) and the one before the previous loop (`A[t - 2]`), 
+which means the operations in the current loop cannot get scheduled until `A[t - 2]` and `A[t - 1]`
+write back.  If we assume that the memory read and write operations both take 1 clock cycle, and 
+the add operation, since it is a combinational logic, can be merged with memory read into one clock 
+cycle, then the RecMII can be estimated as 2. The overall minimum initiation interval is $max(ResMII, 
+RecMII) = 2$.
+
+## Implementation
+I implemented an [LLVM pass](https://github.com/Sibylau/LLVM_10.0_walkthru/tree/master/llvm-pass/loop_II)
+to analyze the innermost loops. The problem is decomposed into two parts: estimate resource-constrained 
+ResMII and data-dependency bounded RecMII.
+
+For ResMII, it is easy to estimate using the formula (1). The pass traverses the loop, calculates 
+the number of independent loads and stores in the code, and divide it by the number of available 
+resources. I applied LLVM global value numbering pass on input programs before my handwritten pass, 

```

Comment:
For completeness: where does the number of available resources come from? Is it specified as a parameter, by the user?

Reply:
Yes, largely. The parameters of available resources come from HLS resource constraints for real application needs, e.g., mapping arrays to1R1W RAMs or dual-port RAMs, or registers, which are also configured in the tool by the user. To use the pass for estimation, the users need to manually set them as MACROs.

Reply:
Cool, thanks.